### ISSUE TITLE
@handle_info ?

### ISSUE THREAD

**Issue by samgaw** [karolsluszniak/phoenix_live_controller#2]
Thanks for sharing! This is a cool lib.

I was trying this out with a simple example but I was wondering if what I'm doing is the right use when it comes to `handle_info` or if there's a better approach that I'm missing?

```elixir
defmodule TestWeb.TimeLive do
  use TestWeb, :live

  @action_mount true
  def index(socket, _params) do
    if connected?(socket), do: :timer.send_interval(1000, self(), :tick)

    now = timestamp()
    assign(socket, time: now)
  end

  def handle_info(:tick, socket), do: {:noreply, timestamp(socket)}

  defp timestamp(), do: NaiveDateTime.local_now()
  defp timestamp(socket), do: assign(socket, time: NaiveDateTime.local_now())

end
```

**Comment by karolsluszniak** [comment]
It's a great question that touches one of the biggest (& most exciting 😄) concerns around LiveController abstraction. I'm very glad it landed here.

Yes, the library doesn't provide any abstractions for GenServer callbacks like handle_info or handle_call. This is because the message may include any term instead of an atom, only in which case it could be mapped into a function the way LiveController does with action mounts or event handlers.

For your case it would be possible to have sth like:

```elixir
defmodule TestWeb.TimeLive do
  use TestWeb, :live

  @action_mount true
  def index(socket, _params) do
    if connected?(socket), do: :timer.send_interval(1000, self(), :tick)

    now = timestamp()
    assign(socket, time: now)
  end

  @info_handler true
  def tick(socket), do: timestamp(socket)

  defp timestamp(), do: NaiveDateTime.local_now()
  defp timestamp(socket), do: assign(socket, time: NaiveDateTime.local_now())
end
```

...but then if you'd have to handle `{:some, :msg, 1}` instead of `:tick` there wouldn't be a simple way of mapping the message to appropriate function name (unless some fancy message matching & routing mechanism would be included in LiveController...). And at this point I don't I think LiveController should cross that river.

This duality between actions/events and the rest of GenServer code is probably why Phoenix.LiveView authors didn't go for this kind of abstraction from the start. But I do believe that this abstraction does have its niche, where it solves issues of extra overhead that LiveView introduces in order to stay closer to GenServer syntax and of unneeded code separation & duplication - [at least it did for me](http://cloudless.studio/articles/51-controller-style-approach-to-liveview-resources). But I'm aware of the other side of the coin - that's why I've included [such extensive explanation of my reasoning in the library docs](https://hexdocs.pm/phoenix_live_controller).

I can see this going one of following ways:
- we stick to two worlds and accept that live controllers have this kind of dual syntax in them
- we separate these two worlds - use LiveController for cases with very little GenServer code & use LiveView for others
- we find a fancy way to extend LiveController to somehow unify the experience for actions/events and for GenServer code, perhaps by learning from libs such as [exactor](https://hex.pm/packages/exactor)
- we find a pattern for organizing code in a way that ensures both worlds coexist within live controller, e.g. delegating handle_info/handle_call to separate (sub)module and leave actions/events & code supporting them in live controller
- we decide that none of the above works & that using this abstraction more often results in trouble than not and drop it altogether

I'll appreciate any feedback on this matter.

**Comment by samgaw** [comment]
Agreed. It may well be niche and not a fit in LiveView proper but as a lib I believe there’s a valid use case for this style of abstraction. Good job so far.

I think if the goal is to simplify projects, the implementation should be as simple as possible. There’s very little benefit in turning complexity into magic. So I a balance of your first & second suggestions could be the best approach:

  - Provide ‘@info_handler’ for situations like refreshing client requests as in your example. 
  - Accept that more complex requirements will require ‘def handle_info’ definitions.

Anything else at this point is likely guessing future needs and going way outside of scope.

**Comment by karolsluszniak** [comment]
You made a convincing point, I gave it a spin and yeah - it seemed like a viable addition for sake of simplicity and consistency. I've merged it in #4 and released v0.3.0 with it. You may want to check out [relevant docs for message handlers](file:///Users/karol/Documents/pro/karolsluszniak/phoenix_live_controller/doc/Phoenix.LiveController.html#module-handling-process-messages).

It supports atoms and tuples with first atom for routing which covers perhaps the most important and useful case of using PubSub like below:

```elixir
defmodule MyAppWeb.ArticleLive do
  use MyAppWeb, :live_controller
  alias Phoenix.PubSub

  @action_handler true
  def show(socket, %{"id" => id}) do
    article = Blog.get_article!(id)
    PubSub.subscribe(MyApp.PubSub, "article:#{article.id}")
    assign(socket, article: Blog.get_article!(id))
  end

  @message_handler true
  def article_update(socket, {_, article}) do
    assign(socket, article: article)
  end

  @event_handler true
  def update(socket = %{assigns: %{article: article}}, %{"article" => article_params}) do
    article = socket.assigns.article

    case Blog.update_article(article, article_params) do
      {:ok, article} ->
        PubSub.broadcast(MyApp.PubSub, "article:#{article.id}", {:article_update, article})

        socket
        |> put_flash(:info, "Article updated successfully.")
        |> push_redirect(to: Routes.article_path(socket, :show, article))

      {:error, %Ecto.Changeset{} = changeset} ->
        assign(socket, article: article, changeset: changeset)
    end
  end
```

Looks quite nice to me - a relevant broadcast & subscribe code all in one place, bringing real-time updates to a resource with just 3 lines of code. Power & pragmatism of LiveView at its best.

**Comment by samgaw** [comment]
I'm going to spend a bit of time today with this but it looks really good. Great work!
